### PR TITLE
shader/validation: Validate trig builtins

### PR DESCRIPTION
**Pull request by ben-clayton**
Fixes: #2658
Fixes: #2659
Fixes: #2660
Fixes: #2661
Fixes: #2662
Fixes: #2663
Fixes: #2664
Fixes: #2665
Fixes: #2666
Fixes: #2670
Fixes: #2671
Fixes: #2672

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+function runTest(
+  t: ShaderValidationTest,
+  builtin: string,
+  expectedResult: boolean,
+  value: number,
+  type: string,
+  stage: string
+) {
+  const enables = requiresShaderF16(type) ? 'enable f16' : '';
```
Comment (jrprice) [inline src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts:41]: Enable directives need a terminating semicolon.

File: src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+/// An array of values ranging from -1 to 2
+const kMinusOneToTwo = [-1.0, -0.9, -0.1, 0.0, 0.1, 0.5, 0.9, 1.0, 1.1, 1.5, 1.9, 2.0] as const;
+
+/// The evaluation stages to test
+const kEvaluationStages = ['constant', 'override'] as const;
+
+/// All floating-point scalar and vector types to test
+const kFloatingPointTypes = [
+  '',
```
Comment (jrprice) [inline src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts:18]: IIUC this is the AbstractFloat case? Presumably this means we aren't testing vectors of AbstractFloats here, but maybe that's OK?

**Review by jrprice — COMMENTED** [review]
Should we be testing that `AbstractInt` implicitly converts to `AbstractFloat` for these builtins here, or will that be covered elsewhere?

Likewise, should we be testing that stuff like `acos(i32(42))` is rejected?

File: src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts (hunk 2 above)
Comment (ben-clayton) [reply src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts:18]: Added abstract vector types to this

File: src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts (hunk 1 above)
Comment (ben-clayton) [reply src/webgpu/shader/validation/expression/call/builtin/trig.spec.ts]: Done

**Comment by ben-clayton** [comment]
> Should we be testing that `AbstractInt` implicitly converts to `AbstractFloat` for these builtins here, or will that be covered elsewhere?

I think this is already covered. JS will omit the `.0` suffix if the number is an integer, which we have cases for.

> Likewise, should we be testing that stuff like `acos(i32(42))` is rejected?

A fair question. But how far do you go with these negative tests? Should we check for 2 arguments? etc.

I can followup with negative tests for bad argument types, if we want to go down this path.

**Comment by ben-clayton** [comment]
Added tests to ensure integer arguments are rejected, and split up the `trig.spec.ts` file into one per builtin.

**Review by jrprice — APPROVED** [review]

**Pull request merged**